File: frontend/src/types/preMarket.ts
/**
 * 盘前扫描类型定义
 * PRD 4.18.0 盘前扫描器
 */

/**
 * 盘前扫描筛选条件
 */
export interface PreMarketScanFilter {
  min_gap: number
  min_premarket_volume: number
  min_volatility: number
  min_liquidity: number
  has_news: boolean | null
  is_earnings_day: boolean | null
}

/**
 * 评分明细
 */
export interface ScoreBreakdown {
  gap: number
  volume: number
  volatility: number
  news: number
  weights: {
    gap: number
    volume: number
    volatility: number
    news: number
  }
}

/**
 * 盘前扫描股票
 */
export interface PreMarketStock {
  symbol: string
  name: string
  gap: number
  gap_direction: 'up' | 'down'
  premarket_price: number
  premarket_volume: number
  premarket_volume_ratio: number
  prev_close: number
  prev_volume: number
  volatility: number
  avg_daily_volume: number
  avg_daily_value: number
  has_news: boolean
  news_headline?: string
  is_earnings_day: boolean
  score: number
  score_breakdown: ScoreBreakdown
}

/**
 * 盘前扫描结果
 */
export interface PreMarketScanResult {
  scan_time: string
  strategy_id: string
  strategy_name: string
  filters_applied: PreMarketScanFilter
  total_matched: number
  stocks: PreMarketStock[]
  ai_suggestion?: string
}

/**
 * 日内监控列表
 */
export interface IntradayWatchlist {
  watchlist_id: string
  user_id: string
  strategy_id: string
  date: string
  symbols: string[]
  created_at: string
  is_confirmed: boolean
}

/**
 * 时间止损配置
 */
export interface TimeStopConfig {
  enabled: boolean
  time: string // HH:mm
}

/**
 * 止盈止损配置
 */
export interface StopLossConfig {
  stop_loss_type: 'atr' | 'fixed' | 'percentage' | 'technical'
  stop_loss_value: number
  take_profit_type: 'atr' | 'fixed' | 'percentage' | 'technical'
  take_profit_value: number
  time_stop_enabled: boolean
  time_stop_time: string
  trailing_stop_enabled: boolean
  trailing_trigger_pct: number
  trailing_distance_pct: number
}

/**
 * 日内交易记录
 */
export interface IntradayTrade {
  time: string
  symbol: string
  side: 'buy' | 'sell'
  quantity: number
  price: number
  stop_loss?: number
  take_profit?: number
  pnl?: number
  is_open: boolean
}

/**
 * 信号状态
 */
export interface SignalStatus {
  type: 'buy' | 'sell' | 'none'
  change: number
  change_pct: number
}

/**
 * 默认筛选条件
 */
export const DEFAULT_FILTERS: PreMarketScanFilter = {
  min_gap: 0.02,
  min_premarket_volume: 2.0,
  min_volatility: 0.03,
  min_liquidity: 5000000,
  has_news: null,
  is_earnings_day: null,
}

/**
 * 默认止盈止损配置
 */
export const DEFAULT_STOP_LOSS_CONFIG: StopLossConfig = {
  stop_loss_type: 'atr',
  stop_loss_value: 1.5,
  take_profit_type: 'atr',
  take_profit_value: 2.5,
  time_stop_enabled: true,
  time_stop_time: '15:55',
  trailing_stop_enabled: false,
  trailing_trigger_pct: 0.5,
  trailing_distance_pct: 0.3,
}

/**
 * 格式化Gap百分比
 */
export function formatGap(gap: number): string {
  const sign = gap >= 0 ? '+' : ''
  return `${sign}${(gap * 100).toFixed(1)}%`
}

/**
 * 获取评分星级
 */
export function getScoreStars(score: number): string {
  const stars = Math.ceil(score / 20)
  return '⭐'.repeat(Math.min(stars, 5))
}

/**
 * 获取评分颜色类
 */
export function getScoreColorClass(score: number): string {
  if (score >= 80) return 'text-green-400'
  if (score >= 60) return 'text-yellow-400'
  return 'text-gray-400'
}

File: backend/app/schemas/pre_market.py
"""
盘前扫描 Schema
PRD 4.18.0 盘前扫描器
"""
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field


class PreMarketScanFilter(BaseModel):
    """盘前扫描筛选条件"""
    min_gap: float = Field(default=0.02, description="最小Gap (默认2%)")
    min_premarket_volume: float = Field(default=2.0, description="盘前成交量倍数 (默认2倍日均)")
    min_volatility: float = Field(default=0.03, description="最小昨日波动率 (默认3%)")
    min_liquidity: float = Field(default=5000000, description="最小流动性 (默认$5M/日)")
    has_news: Optional[bool] = Field(None, description="是否有新闻")
    is_earnings_day: Optional[bool] = Field(None, description="是否财报日")


class ScoreBreakdown(BaseModel):
    """评分明细"""
    gap: float
    volume: float
    volatility: float
    news: float
    weights: dict = {"gap": 0.3, "volume": 0.3, "volatility": 0.2, "news": 1.0}


class PreMarketStock(BaseModel):
    """盘前扫描股票"""
    symbol: str
    name: str

    # 盘前数据
    gap: float  # 开盘跳空 (%)
    gap_direction: str  # 'up' | 'down'
    premarket_price: float  # 盘前价格
    premarket_volume: int  # 盘前成交量
    premarket_volume_ratio: float  # 相对日均量倍数

    # 昨日数据
    prev_close: float
    prev_volume: int
    volatility: float  # 昨日波动率 (ATR%)

    # 流动性
    avg_daily_volume: int
    avg_daily_value: float  # 日均成交额

    # 新闻/事件
    has_news: bool
    news_headline: Optional[str] = None
    is_earnings_day: bool = False

    # 评分
    score: float  # 策略评分 0-100
    score_breakdown: ScoreBreakdown


class PreMarketScanResult(BaseModel):
    """盘前扫描结果"""
    scan_time: datetime
    strategy_id: str
    strategy_name: str

    filters_applied: PreMarketScanFilter
    total_matched: int
    stocks: list[PreMarketStock]

    # AI建议
    ai_suggestion: Optional[str] = None


class IntradayWatchlist(BaseModel):
    """日内交易监控列表"""
    watchlist_id: str
    user_id: str
    strategy_id: str
    date: date
    symbols: list[str]
    created_at: datetime
    is_confirmed: bool = False


class CreateWatchlistRequest(BaseModel):
    """创建监控列表请求"""
    strategy_id: str
    symbols: list[str] = Field(..., max_length=20)


class TimeStopConfig(BaseModel):
    """时间止损配置"""
    enabled: bool = True
    time: str = "15:55"  # HH:mm format, 收盘前5分钟


class StopLossConfig(BaseModel):
    """止盈止损配置"""
    # 止损设置
    stop_loss_type: str = "atr"  # 'atr' | 'fixed' | 'percentage' | 'technical'
    stop_loss_value: float = 1.5

    # 止盈设置
    take_profit_type: str = "atr"
    take_profit_value: float = 2.5

    # 时间止损 (日内专属)
    time_stop_enabled: bool = True
    time_stop_time: str = "15:55"

    # 移动止损
    trailing_stop_enabled: bool = False
    trailing_trigger_pct: float = 0.5
    trailing_distance_pct: float = 0.3


class IntradayPosition(BaseModel):
    """日内持仓"""
    position_id: str
    user_id: str
    account_id: str
    strategy_id: str

    symbol: str
    quantity: int
    entry_price: float
    current_price: float
    avg_cost: float

    pnl: float
    pnl_pct: float

    stop_loss: Optional[float] = None
    take_profit: Optional[float] = None
    time_stop_config: Optional[TimeStopConfig] = None

    entry_time: datetime
    is_intraday: bool = True
